code-studio/src/js/clientState.js

```diff
@@ -1,34 +1,160 @@
- *       combination of cookies and HTML5 web storage.
+ * @file A module for accessing persistent per-user client state.
+ *
+ * Client state for signed-in users persists even across logins.
+ * Client-state for signed-out users is per-tab and goes aways when the tab
+ * is closed.
+ *
+ * Before getting or setting client state, the developer must call
+ * setCurrentUserKey with a unique key for the current user, or setAnonymousUser
+ * to use temporary state for an anonymous user.
+ *
+ * Client state for logged in users is "reasonably" durable but not quite hard
+ * state-- the client state module garbage-collects the oldest client state
+ * if the total size of the client state exceeds the 5MB browser limit. In
+ * practice this limit will rarely be reached, but in this case this does
+ * occur the database typically has the same state and can be used to refill
+ * the client state cache.
+ *
+ * Implementation notes:
+ * We use the 'lscache' npm module to divide local storage into per-user buckets
+ * and to expire the least recently created storage slots.
+ *
+ * The current user key for is kept in the 'user_key' slot in localStorage. For
+ * unauthenticated users, where the user_key is the empty string, a second
+ * 'anon_user_key' slot in *session* storage is used to store a randomly
+ * generated temporary key. By using session storage, we ensure that each tab
+ * uses its own key and has its own client state bucket. The temporary
+ * localStorage for anonymous user keys is cleared when setCurrentUserKey is
+ * called for a signed in user.
+ *
+ * The code that uses this module typically merges the client state with the
+ * state from the server (if available) so it captures both progress on the
+ * current machine and progress by the same user on other machines.
  */
+
 'use strict';
 
-module.exports = function (sessionStorage, $) {
+var lscache = require('lscache');
+
+module.exports = function () {
 
 var clientState = {};
 
 /**
- * Number of days before client state cookie expires.
- * @type {number}
- * @private
+ * Max number of minutes before client state expires, current 10 years.
+ * @private {number}
  */
-clientState.EXPIRY_DAYS = 365;
+clientState.EXPIRY_MINUTES = 365 * 10 * 24 * 60;
 
 /**
- * Maximum number of lines of code that can be stored in the cookie
- * @type {number}
- * @private
+ * Maximum number of lines of code that can be added at once.
+ * @private {number}
  */
 var MAX_LINES_TO_SAVE = 1000;
 
-var COOKIE_OPTIONS = {expires: clientState.EXPIRY_DAYS, path: '/'};
+/**
+ * A special key indicating that the user is not signed in.
+ * @const {string}
+ */
+var ANON_USER_KEY = '';
+
+/**
+ * Returns the storage key for the current user.
+ * @return {string}
+ */
+function getStorageBucketForCurrentUser() {
+  var userBucket = localStorage.getItem('user_key');
+  if (userBucket == ANON_USER_KEY) {
+    userBucket = sessionStorage.getItem('anon_key');
```

Comment:
I sort of prefer using a session cookie here, so the anon progress persists across tabs.  Otherwise it's weird that opening the course overview page in a new tab doesn't show your progress if you're logged out.  Thoughts?